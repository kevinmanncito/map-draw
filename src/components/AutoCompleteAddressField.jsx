import React from 'react'
import PropTypes from 'prop-types'

class AutoCompleteAddressField extends React.Component {
  static propTypes = {
    placeholder: PropTypes.string,
    coordsFound: PropTypes.func.isRequired,
  }

  autocomplete = undefined;

  componentDidMount () {
    this.addressFound = this.addressFound.bind(this)

    // Create the autocomplete object, restricting the search to geographical
    // location types.
    this.autocomplete = new google.maps.places.Autocomplete(
      (document.getElementById('autocomplete')),
      {types: ['geocode']}
    )

    // When the user selects an address from the dropdown, populate the address
    // fields in the form.
    this.autocomplete.addListener('place_changed', this.addressFound)
  }

  addressFound (address) {
    const place = this.autocomplete.getPlace()
    if (place.geometry) {
      const lat = place.geometry.location.lat()
      const lon = place.geometry.location.lng()
      this.props.coordsFound({
        lat: lat,
        lon: lon,
      })
    }
  }

  render () {
    return (
      <div>
        <input
          id='autocomplete'
          type='text'
          className='autocomplete-field'
          placeholder={this.props.placeholder} />
      </div>
    )
  }
}

export default AutoCompleteAddressField
