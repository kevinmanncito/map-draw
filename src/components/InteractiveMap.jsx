import React from 'react'
import PropTypes from 'prop-types'

class InteractiveMap extends React.Component {
  static propTypes = {
    areaCalculated: PropTypes.func.isRequired,
    coords: PropTypes.shape({
      lat: PropTypes.number.isRequired,
      lon: PropTypes.number.isRequired,
    }).isRequired,
  }
  map = undefined
  draw = undefined

  componentWillReceiveProps (props) {
    this.map.setCenter([props.coords.lon, props.coords.lat])
    this.map.setZoom(18)
  }

  componentDidMount () {
    this.updateArea = this.updateArea.bind(this)

    mapboxgl.accessToken = 'pk.eyJ1Ijoia2V2aW5ybWFubiIsImEiOiJBTWUyS05zIn0.ZPXjYLQs4QIJiqp85cfI0w'
    this.map = new mapboxgl.Map({
      container: 'map', // container id
      style: 'mapbox://styles/kevinrmann/cjaj68e7dakuc2slmhkk3bav6', // hosted style id
      center: [this.props.coords.lat, this.props.coords.lon],
      zoom: 4, // starting zoom
    })

    this.draw = new MapboxDraw({
      displayControlsDefault: false,
      controls: {
        polygon: true,
        trash: true,
      },
    })
    this.map.addControl(this.draw)

    this.map.on('draw.create', this.updateArea)
    this.map.on('draw.delete', this.updateArea)
    this.map.on('draw.update', this.updateArea)
  }

  updateArea (e) {
    const data = this.draw.getAll()
    if (data.features.length > 0) {
      const area = turf.area(data)
      this.props.areaCalculated(area)
    } else {
      this.props.areaCalculated(undefined)
    }
  }

  render () {
    return (
      <div id='map' className='content' />
    )
  }
}

export default InteractiveMap
