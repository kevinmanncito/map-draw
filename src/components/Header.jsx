import React from 'react'
import PropTypes from 'prop-types'

import AutoCompleteAddressField from './AutoCompleteAddressField'

class Header extends React.Component {
  static propTypes = {
    coordsFound: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired,
  }

  render () {
    return (
      <header className='header'>
        <h1>{this.props.title}</h1>
        <AutoCompleteAddressField
          coordsFound={(coords) => this.props.coordsFound(coords)}
          placeholder='Enter address' />
      </header>
    )
  }
}

export default Header
