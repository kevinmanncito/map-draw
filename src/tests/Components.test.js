import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import sinon from 'sinon'

import Header from '../components/Header'
import AutoCompleteAddressField from '../components/AutoCompleteAddressField'
import InteractiveMap from '../components/InteractiveMap'

configure({ adapter: new Adapter() })

describe('Components render correctly', () => {
  it('Header renders correctly', () => {
    const wrapper = shallow(
      <Header title={'Test Title'} coordsFound={jest.fn()} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('AutoCompleteAddressField renders correctly', () => {
    class AutocompleteStub {
      addListener () {
        return null
      }
    }
    let googleStub = sinon.stub()
    googleStub.maps = {
      places: {
        Autocomplete: AutocompleteStub,
      },
    }
    global.google = googleStub

    const wrapper = shallow(
      <AutoCompleteAddressField coordsFound={jest.fn()} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('InteractiveMap renders correctly', () => {
    class mapStub {
      addControl () {
        return null
      }
      on () {
        return null
      }
    }
    global.mapboxgl = {
      Map: mapStub,
    }
    global.MapboxDraw = sinon.stub()
    const wrapper = shallow(
      <InteractiveMap
        areaCalculated={jest.fn()}
        coords={{lat: -95.587, lon: 41.313}} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
