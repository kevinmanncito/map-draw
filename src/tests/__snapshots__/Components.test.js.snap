// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components render correctly AutoCompleteAddressField renders correctly 1`] = `
<div>
  <input
    className="autocomplete-field"
    id="autocomplete"
    type="text"
  />
</div>
`;

exports[`Components render correctly Header renders correctly 1`] = `
<header
  className="header"
>
  <h1>
    Test Title
  </h1>
  <AutoCompleteAddressField
    coordsFound={[Function]}
    placeholder="Enter address"
  />
</header>
`;

exports[`Components render correctly InteractiveMap renders correctly 1`] = `
<div
  className="content"
  id="map"
/>
`;
